source/digits_hits/src/GateRestrictedVolume.cc

```diff
@@ -0,0 +1,67 @@
+/** ----------------------
+  Copyright (C): OpenGATE Collaboration
+  This software is distributed under the terms
+  of the GNU Lesser General  Public Licence (LGPL)
+  See LICENSE.md for further details
+  ----------------------*/
+
+#include "GateRestrictedVolume.hh"
+
+#include "G4UnitsTable.hh"
+
+#include "GateRestrictedVolumeMessenger.hh"
+#include "GateTools.hh"
+
+GateRestrictedVolume::GateRestrictedVolume(GatePulseProcessorChain* itsChain,
+                                       const G4String& itsName)
+  : GateVPulseProcessor(itsChain, itsName)
+{
+  m_messenger = new GateRestrictedVolumeMessenger(this);
+}
+
+GateRestrictedVolume::~GateRestrictedVolume()
+{
+  delete m_messenger;
+}
+
+void GateRestrictedVolume::ProcessOnePulse(const GatePulse* inputPulse, GatePulseList& outputPulseList)
+{
+  if (!inputPulse) {
+    if (nVerboseLevel>1)
+      G4cout << "[GateRestrictedVolume::ProcessOnePulse]: input pulse was null -> nothing to do\n\n";
+    return;
+  }
+  
+  if (((inputPulse->GetVolumeID()).GetBottomCreator())->GetObjectName() == "") {
+    if (nVerboseLevel>1)
+      G4cout << "[GateRestrictedVolume::ProcessOneHit]: volume is null for " << inputPulse << " -> pulse ignored\n\n";
+    return;
+  }
+  GatePulse* outputPulse = new GatePulse(*inputPulse);
+  // Reject mode and the volume is the one specified
+  if (((inputPulse->GetVolumeID()).GetBottomCreator())->GetObjectName() != m_restrictedVolume && m_restrictionMode == "reject")
+  {
+    outputPulseList.push_back(outputPulse);
```

Comment:
Check if this vector delete all its elements somewhere, otherwise we have memory leak here.